### PR TITLE
Extract doc_state into several monoids

### PR DESCRIPTION
**Pull request by V-FEXrt**
Splits doc_state into several monoids, enabling easier composition of other bits of data.
 
I started down the template path but it turned out to be quite combursome and prone so I'm hoping to put that off for now until it becomes more important.

max_width also isn't implemented, happy to throw it in but since we weren't using it I didn't put in the effort to pin it down

### CODE REVIEW

**Comment by JakeSiFive** [comment]
monoids not monads...

monoids = "things you can put together with +, string concat is the best example but integers work also"
monads = "...oh boy....there's like a million different articles trying to explain this one intuitively...I didn't understand any of them until one day I understood monads and then they all seemed so obvious"

There's a famous saying/joke thing that people say like "monads are just monoids in the category of endofunctors" which legit hardly anyone, even people that know how to use monads, understands. It's the category theoretic definition of a monad but it's not very helpful from a programmers perspective. You have to extract some usefulness out of that definition via some math legwork before you can get to the programmers definition.

**Comment by V-FEXrt** [comment]
lol FWIW I had it right the first time then second guessed myself

File: tools/wake-format/formatter.h [hunk 1]
```diff
@@ -189,9 +189,9 @@ struct IfElseAction {
 
   ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
     if (predicate(builder, ctx, node)) {
-      builder.append(if_formatter.compose(ctx.sub(builder), node));
+      builder.append(if_formatter.compose(ctx + builder, node));
```
Comment (JakeSiFive) [inline tools/wake-format/formatter.h:192]: Heh...this is clever but I think I'm still in favor of using `sub` for this. The types are different here and this hides an important point

File: tools/wake-format/formatter.h [hunk 2]
```diff
@@ -35,34 +35,34 @@
...
 
-  ctx_t nest() {
+  ctx_t nest() const {
     ctx_t copy = *this;
     copy.nest_level++;
     return copy;
   }
 
-  ctx_t sub(const wcl::doc_builder& builder) {
+  ctx_t operator+(const wcl::doc_builder& builder) const {
```
Comment (JakeSiFive) [inline tools/wake-format/formatter.h:47]: nit: I think we should go back to `sub` 

But this 100% should be `const`...sorry I missed that before!

File: src/wcl/doc_state.h [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+  static last_ws_count_state inject(size_t size, utf8proc_int32_t codepoint) {
+    if (codepoint == '\n') return last_ws_count_state{true, 0};
+    return last_ws_count_state{false, codepoint == ' ' ? 1u : 0u};
+  }
+
+  friend std::hash<last_ws_count_state>;
+};
+
+class doc_state {
```
[the remark this answers is not in the document]
Comment (V-FEXrt) [reply src/wcl/doc_state.h:191]: Making this a template type turned out to be exceptionally painful and I don't think we should do it right now. Without doing that, moving this to formatter.h would create a cyclic dep IIRC

File: src/wcl/doc_state.h [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+namespace wcl {
+template <class State>
+State from_string(const std::string& str) {
+  State out = State::identity();
+
+  const utf8proc_uint8_t* iter = reinterpret_cast<const utf8proc_uint8_t*>(str.c_str());
+  const utf8proc_uint8_t* iter_end = iter + str.size();
+  while (iter < iter_end) {
+    utf8proc_int32_t codepoint;
+    size_t size = utf8proc_iterate(iter, -1, &codepoint);
```
[the remark this answers is not in the document]
Comment (V-FEXrt) [reply src/wcl/doc_state.h:36]: I agree and that is what I did originally, but I couldn't find any function from codepoint->byte_size. Happy to change it if you can find one.

File: src/wcl/doc_state.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (JakeSiFive) [reply src/wcl/doc_state.h:191]: You're 100% right about the cyclic dep...hmm ok I'm happy to punt on making it all templated I suppose. doc_state is feeling ad hoc but knowing that we can escape the ad hoc nature seems fine to me.

[the remark this answers is not in the document]
Comment (V-FEXrt) [reply src/wcl/doc_state.h:191]: yeah, agreed. we can review later if it gets even more unruly.

or we can leave as a starter project for the next wake C++ dev ;)

File: src/wcl/doc_state.h [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+struct empty_state {
+  empty_state operator+(empty_state other) const { return empty_state{}; }
+
+  bool operator==(const empty_state& other) const { return true; }
+
+  static empty_state identity() { return empty_state{}; }
+
+  static empty_state inject(size_t size, utf8proc_int32_t codepoint) { return empty_state{}; }
```
[the remark this answers is not in the document]
Comment (V-FEXrt) [reply src/wcl/doc_state.h:54]: see above

File: src/wcl/doc_state.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (JakeSiFive) [reply src/wcl/doc_state.h:36]: ok just please use "byte_size" everywhere to make it clear. I read it as width at first (which was an incorrect reading)

**Review by JakeSiFive — APPROVED** [review]

**Pull request merged**